### ISSUE TITLE
StringBuilder Pool Crashes Site

### ISSUE THREAD

**Issue by deanmarcussen** [OrchardCMS/OrchardCore#9179]
Somethings going a little weird with the `StringBuilder` pool, and the `ViewBufferTextWriterContent`

About to head into a meeting, but the current repo is just to refresh the agency landing page a few times.

Sooner or later, there's an error in the `StringBuilderPool.GetInstance() `

```
2021-04-15 14:07:39.1363|INFO|Microsoft.Hosting.Lifetime|Now listening on: http://localhost:5000
2021-04-15 14:07:39.1471|INFO|Microsoft.Hosting.Lifetime|Now listening on: https://localhost:5001
2021-04-15 14:07:39.1476|INFO|Microsoft.Hosting.Lifetime|Application started. Press Ctrl+C to shut down.
2021-04-15 14:07:39.1476|INFO|Microsoft.Hosting.Lifetime|Hosting environment: Development
2021-04-15 14:07:39.1476|INFO|Microsoft.Hosting.Lifetime|Content root path: /Users/deanmarcussen/Projects/OrchardCore/src/OrchardCore.Cms.Web
Process terminated. Assertion failed.
   at OrchardCore.DisplayManagement.StringBuilderPool.GetInstance() in /Users/deanmarcussen/Projects/OrchardCore/src/OrchardCore/OrchardCore.DisplayManagement/ObjectPool.cs:line 240
   at OrchardCore.DisplayManagement.Liquid.ViewBufferTextWriterContent.AllocateBuilder() in /Users/deanmarcussen/Projects/OrchardCore/src/OrchardCore/OrchardCore.DisplayManagement.Liquid/ViewBufferTextWriterContent.cs:line 61
   at OrchardCore.DisplayManagement.Liquid.ViewBufferTextWriterContent.Write(String value) in /Users/deanmarcussen/Projects/OrchardCore/src/OrchardCore/OrchardCore.DisplayManagement.Liquid/ViewBufferTextWriterContent.cs:line 89
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

When this happens, the instance coming back from the string builder has a value, i.e.

<img width="753" alt="Screenshot 2021-04-15 at 14 04 57" src="https://user-images.githubusercontent.com/13782679/114875051-d1a89700-9df4-11eb-97b7-510f74f44c13.png">

Need to change branches and make sure it's not something I've done.

Latest dev

**Comment by deanmarcussen** [comment]
so a few things going on here.

definite issues with fluid 2.0.4

one render
<img width="803" alt="Screenshot 2021-04-15 at 15 31 59" src="https://user-images.githubusercontent.com/13782679/114893572-bdb96100-9e05-11eb-87c3-8f8120d79341.png">


next render

<img width="798" alt="Screenshot 2021-04-15 at 16 14 10" src="https://user-images.githubusercontent.com/13782679/114893601-c4e06f00-9e05-11eb-9014-24112b0f7273.png">

**Comment by jtkech** [comment]
Just after a quick look so not sure, but i think it's because the execution of some tasks in parallel, here e.g. https://github.com/lahma/fluid/blob/cbf7f887ab25c5bea4fe14791aaff770b1e09899/Fluid/Ast/ElseIfStatement.cs#L24 and that write to the writer that may be our `ViewBufferTextWriterContent`.

**Comment by jtkech** [comment]
**Update**: Talked too quick ;)

But @sebastienros found the issue !!!

**Comment by hishamco** [comment]
@jtkech is this have been resolved by the Async methods added by Seb?

**Comment by jptissot** [comment]
I think so. I tested on my solution using the nuget pacakges and everything seems fine now.

**Comment by jtkech** [comment]
Yes i could repro the issue too and seems to be fixed by the PR of @sebastienros 

But now @sebastienros has another issue, but that i can't repro on my side, so for now i can't help ;)

**Comment by jptissot** [comment]
I tested my site with the latest nuget packages and it seems to fix the issue.

**Comment by lahma** [comment]
What a fun issue that was, glad that Sébastien figured out the root cause. Not only fixing the issue it might even improve performance on top of the Fluid performance tweaks. The default (and weird) implementation that `TextWriter` had seemed quite sub-optimal for the use case.

**Comment by deanmarcussen** [comment]
Fixed by https://github.com/OrchardCMS/OrchardCore/pull/9180
